File: python/dnn_control/simulator.py
class Simulator:

    '''
        This class implements the interaction between stepwise integration of the ode system,
        artificial sensor data generation, control polling, and result logging.
    '''

    def __init__(self, asteroid, sensor_simulator, spacecraft_controller, control_frequency, perturbation_noise):
        """
        Constructor. The spacecraft state is defined as follows:
        0 : position x
        1 : position y
        2 : position z
        3 : velocity x
        4 : velocity y
        5 : velocity z
        6 : mass

        The simulator expects the controller to return the following vector when calling get_thrust():
        0 : thrust x
        1 : thrust y
        2 : thrust z
        """

        from odesystem import ODESystem
        from scipy.integrate import ode

        self._sensor_simulator = sensor_simulator
        self._spacecraft_controller = spacecraft_controller
        self._control_frequency = float(control_frequency)
        self._control_interval = 1.0 / self._control_frequency
        self._perturbation_noise = perturbation_noise

        self._system = ODESystem(asteroid)
        self._integrator = ode(self._system.dynamics)
        self._integrator.set_integrator("lsoda")
        self._log_states = []

    # Simulates random noise in the dynamical system (which is fed into the ODE system "system_")
    def _simulate_perturbations(self):
        from numpy import random

        mass = self._integrator.y[6]
        return [mass * random.normal(0.0, self._perturbation_noise),
                mass * random.normal(0.0, self._perturbation_noise),
                mass * random.normal(0.0, self._perturbation_noise)]

    # Before running the simulator, it has to be initialized with a spacecraft configuration
    def init_spacecraft(self, position, velocity, mass, specific_impulse):
        from constants import EARTH_ACCELERATION

        # Transform position, velocity and mass to state
        self._integrator.set_initial_value(position + velocity + [mass], 0.0)

        # Cache g * Isp
        self._system.coef_earth_acceleration_mul_specific_impulse = 1.0 / (specific_impulse * EARTH_ACCELERATION)

    # Initialize spacecraft only by specifying the spacecraft's specific impulse (useful for the function next_state)
    def init_spacecraft_specific_impulse(self, specific_impulse):
        from constants import EARTH_ACCELERATION

        # Cache g * Isp
        self._system.coef_earth_acceleration_mul_specific_impulse = 1.0 / (specific_impulse * EARTH_ACCELERATION)

    # Implements F: S x A x T -> S : F(s,a,t) = s' (Useful for RL?)
    def next_state(self, state, thrust, time):
        # Assign state
        self._integrator.set_initial_value(state, time)
        self._system.thrust = thrust

        # Simulate perturbations, directly write it to the ode system
        self._system.perturbations_acceleration = self._simulate_perturbations()

        # Integrate for one time step _control_interval
        self._integrator.integrate(time + self._control_interval)

        # Extract new state out of system
        return self._integrator.y[:]

    # Simulates the system for time "time". Logs the states if "log_data" is enabled,
    # Returns the number of iterations the simulator made to get to the specified time.
    def run(self, time, log_data=False):
        from numpy import zeros

        dt = self._control_interval
        iterations = int(time / dt)

        if log_data:
            self._log_states = zeros([iterations, 2, 3])

        current_time = 0.0
        # Stepwise integration for "iterations" iterations
        try:
            for iteration in xrange(iterations):
                if log_data:
                    # Log position and height, if enabled
                    position = list(self._integrator.y[0:3])
                    surface_position = self._system.asteroid.nearest_point_on_surface_to_position(position)[0]
                    for i in [0, 1, 2]:
                        self._log_states[iteration][0][i] = position[i]
                        self._log_states[iteration][1][i] = position[i] - surface_position[i]

                # Simulate perturbations, directly write it to the ode system
                self._system.perturbations_acceleration = self._simulate_perturbations()

                # Simulate sensor data
                sensor_data = self._sensor_simulator.simulate(self._integrator.y,
                                                             self._system.perturbations_acceleration,
                                                             current_time)

                # Poll controller for control thrust, write it to the ode system
                self._system.thrust = self._spacecraft_controller.get_thrust_for_sensor_data(sensor_data)

                # Integrate the system for _control_interval time
                self._integrator.integrate(current_time + dt)
                if not self._integrator.successful():
                    print("Something BAD happened...")
                    exit(1)

                current_time += dt

                # Check if spacecraft is out of fuel
                if self._integrator.y[6] <= 0.0:
                    print("The spacecraft is out of fuel.")
                    break

        except RuntimeError as error:
            print("The spacecraft crashed into the asteroid's surface.")

        return current_time

    # Writes the logged data from Run to the file given by "path_to_file"
    def flush_log_to_file(self, path_to_file):
        from os import remove

        try:
            remove(path_to_file)
        except OSError:
            pass

        log_file = open(path_to_file, "a")
        log_file.write("{0},\t{1},\t{2},\t{3}\n".format(self._system.asteroid.get_semi_axis(0),
                                                  self._system.asteroid.get_semi_axis(1),
                                                  self._system.asteroid.get_semi_axis(2),
                                                  self._control_frequency))
        for state in self._log_states:
            log_file.write("{0},\t{1},\t{2},\t{3},\t{4},\t{5}\n".format(state[0][0], state[0][1], state[0][2],
                                                            state[1][0], state[1][1], state[1][2]))
        log_file.close()
        self._log_states = []

File: cpp/dnn_control/simulator.cpp
#include "simulator.h"
#include "constants.h"
#include "odeint.h"

Simulator::Simulator(Asteroid &asteroid, SensorSimulator *sensor_simulator, SpacecraftController *spacecraft_controller, const double &control_frequency, const double &perturbation_noise) :
    sensor_simulator_(sensor_simulator), spacecraft_controller_(spacecraft_controller), system_(ODESystem(asteroid)), normal_distribution_(boost::mt19937(time(0)),boost::normal_distribution<>(0.0, perturbation_noise)) {
    control_frequency_ = control_frequency;
    control_interval_ = 1.0 / control_frequency;
}

Simulator::~Simulator() {
    delete sensor_simulator_;
    delete spacecraft_controller_;
}

void Simulator::InitSpacecraft(const Vector3D &position, const Vector3D &velocity, const double &mass, const double &specific_impulse) {
    // Transform position, velocity and mass to state
    for (int i = 0; i < 3; ++i) {
        system_.state_[i] = position[i];
        system_.state_[3+i] = velocity[i];
    }
    system_.state_[6] = mass;

    // Cache g * Isp
    system_.coef_earth_acceleration_mul_specific_impulse_ = 1.0 / (specific_impulse * k_earth_acceleration);
}

void Simulator::InitSpacecraftSpecificImpulse(const double &specific_impulse) {
    // Cache g * Isp
    system_.coef_earth_acceleration_mul_specific_impulse_ = 1.0 / (specific_impulse * k_earth_acceleration);
}

void Simulator::NextState(const State &state, const Vector3D &thrust, const double &time, State &next_state) {
    using namespace boost::numeric::odeint;

    // Assign state
    for (int i = 0; i < 3; ++i) {
        system_.state_[i] = state[i];
        system_.state_[3+i] = state[3+i];
        system_.thrust_[i] = thrust[i];
    }
    system_.state_[6] = state[6];

    // Simulate perturbations, directly write it to the ode system
    SimulatePerturbations(system_.perturbations_acceleration_);

    // Integrate for one time step control_interval_
    runge_kutta4<State> integrator;
    integrator.do_step(system_, system_.state_, time, control_interval_);

    // Extract new state out of system
    for (int i = 0; i < 7; ++i) {
        next_state[i] = system_.state_[i];
    }
}

boost::tuple<double, std::vector<Vector3D>, std::vector<SensorData> > Simulator::Run(const double &time, const bool &log_sensor_data) {
    using namespace boost::numeric::odeint;

    std::vector<SensorData> logged_sensor_data;
    std::vector<Vector3D> logged_positions;

    const int sensor_dimensions = sensor_simulator_->Dimensions();
    if (sensor_dimensions != spacecraft_controller_->Dimensions()) {
        std::cout << "Warning: sensor simulator and spacecraft controller have different output/input dimensions." << std::endl;
    }
    const double dt = control_interval_;
    const int iterations = time / dt;

    logged_positions = std::vector<Vector3D>(iterations);

    if (log_sensor_data) {
        logged_sensor_data = std::vector<SensorData>(iterations, std::vector<double>(sensor_dimensions,0));
    }

    runge_kutta4<State> integrator;
    double current_time = 0.0;
    SensorData sensor_data(sensor_dimensions,0);

    // Stepwise integration for "iterations" iterations
    try {
        for(int iteration = 0; iteration < iterations; ++iteration) {
            Vector3D &iter_pos = logged_positions.at(iteration);
            iter_pos[0] = system_.state_[0];
            iter_pos[1] = system_.state_[1];
            iter_pos[2] = system_.state_[2];

            // Simulate perturbations, directly write it to the ode system
            SimulatePerturbations(system_.perturbations_acceleration_);

            // Simulate sensor data
            sensor_simulator_->Simulate(system_.state_, system_.perturbations_acceleration_, current_time, sensor_data);

            if (log_sensor_data) {
                // Log sensor data, if enabled
                SensorData &data = logged_sensor_data.at(iteration);
                for (int i = 0; i < sensor_dimensions; ++i) {
                    data[i] = sensor_data[i];
                }
            }

            // Poll controller for control thrust, write it to the ode system
            spacecraft_controller_->GetThrustForSensorData(sensor_data, system_.thrust_);

            // Integrate the system for control_interval_ time
            integrator.do_step(system_, system_.state_, current_time, dt);

            current_time += dt;

            // Check if spacecraft is out of fuel
            if (system_.state_[6] <= 0.0) {
                std::cout << "The spacecraft is out of fuel." << std::endl;
                break;
            }
        }
    } catch (boost::exception const &exception) {
        std::cout << "The spacecraft crashed into the asteroid's surface." << std::endl;
    }

    return boost::make_tuple(current_time, logged_positions, logged_sensor_data);
}

void Simulator::SimulatePerturbations(Vector3D &perturbations) {
    double mass = system_.state_[6];
    for(int i = 0; i < 3; ++i) {
        perturbations[i] = mass * normal_distribution_();
    }
}
